controllers/storagecluster/consumer.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// onBoardConsumer makes an API call to the provider cluster for onboarding
+func onBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	// TODO: think about naming convention
+	response, err := r.providerClient.OnboardConsumer("my-ocs", getExternalStorageRequestedCapacity(instance))
+	err = r.handleOnboardConsumerErrors(instance, err)
+	if err != nil {
+		return err
+	}
+
+	if response.StorageConsumerUUID == "" || response.GrantedCapacity == "" {
+		err = fmt.Errorf("External-OCS:OnboardConsumer:response is empty")
+		r.Log.Error(err, "empty response")
+		return err
+	}
+
+	instance.Status.ExternalStorage.ConsumerID = response.StorageConsumerUUID
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return nil
+}
+
+// offBoardConsumer makes an API call to the provider cluster for offboarding
+func offBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	_, err := r.providerClient.OffboardConsumer(instance.Status.ExternalStorage.ConsumerID)
+	return r.handleOffboardConsumerErrors(instance, err)
+}
+
+// upscaleOrDownscaleConsumer makes an API call to the provider cluster to update the capacity
+func updateConsumerCapacity(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	response, err := r.providerClient.UpdateCapacity(
+		instance.Status.ExternalStorage.ConsumerID,
+		getExternalStorageRequestedCapacity(instance))
+	err = r.handleUpdateConsumerCapacityErrors(instance, err)
+	if err != nil {
+		return err
+	}
+
+	if response.GrantedCapacity == "" {
+		err = fmt.Errorf("External-OCS:UpdateCapacity:response is empty")
+		r.Log.Error(err, "empty response")
+		return err
+	}
+
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return nil
+}
+
+// getExternalConfigFromProvider makes an API call to the provider cluster for json blob
+// and save it in a global var (externalOCSResources) to use the same json again and again
+// It also skips the API call if json is already available
+func getExternalConfigFromProvider(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) ([]ExternalResource, error) {
+
+	if externalOCSResources[instance.UID] != nil {
+		return externalOCSResources[instance.UID], nil
+	}
+
+	config, err := r.providerClient.GetStorageConfig(instance.Status.ExternalStorage.ConsumerID)
+	err = r.handleGetStorageConfigErrors(instance, err)
+	if err != nil {
+		return nil, err
+	}
+
+	var externalResources []ExternalResource
+	err = json.Unmarshal(config.Data, &externalResources)
+	if err != nil {
+		return nil, err
+	}
+
+	externalOCSResources[instance.UID] = externalResources
+
+	return externalOCSResources[instance.UID], nil
+}
+
+// getConsumerClient returns the *providerClient.ConsumerClient
+func getProviderClient(instance *ocsv1.StorageCluster) (*grpc.ClientConn, *providerClient.OCSProviderClient, error) {
```

Comment:
shouldn't it called newProviderClient or CreateProviderClient?

Reply:
Why is not the grpc connection part of the providerClient?

Reply:
> shouldn't it called newProviderClient or CreateProviderClient?

This can be done

> Why is grpc connection not part of the providerClient?

@sp98 Can you pls answer this one?